### ISSUE TITLE
Possible Error for tf Gradient on mean

### ISSUE THREAD

**Issue by JaeDukSeo** [tensorflow/tensorflow#23665]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): window 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version: python 3
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with
python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
b'v1.8.0-0-g93bc2e2072' 1.8.0

**Describe the current behavior**
When taking the gradient respect to a variable and reduce mean the outcome should be 
1 - (1/(# of reduced dimension)) however it does not happen. 

**Describe the expected behavior**
There are a different blog post and resources on how to perform backpropagation for batch norm. 
Centering the data is a key part of any norm, and it seems like auto differentiation is getting the value wrong? Or am I just going crazy. I thought the gradient respect to the input variable should be
1 - 1/(10*94*94) - since 10 is dim - 94 is dim 1 and 2. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
import tensorflow as tf
testing_tensor   = tf.placeholder(shape=[10,94,94,32],dtype=tf.float64)
testing_output   = testing_tensor - tf.reduce_mean(testing_tensor,(0,1,2),keep_dims=True) 
what_grad_auto   = tf.gradients(xs=testing_tensor,ys=testing_output)

![image](https://user-images.githubusercontent.com/22832406/48318218-c9682400-e5ca-11e8-8abd-46e660ada701.png)

**Comment by JaeDukSeo** [comment]
Sorry for my horrible explanation, I have the code in my GitHub repo. Or my math might just be wrong.
Different blog post such as https://medium.com/samkirkiles/spatial-batchnorm-backprop-implementation-notes-8ccde1ac62a2 describes how to do a back prop for batch norm as well

**Comment by ppwwyyxx** [comment]
> Or my math might just be wrong.

Indeed.
The code is working correctly.

**Comment by JaeDukSeo** [comment]
I get that but can you show me how the derivative work out? (assuming you know how to). I think when the mean subtraction layer is c and original input is x (10,94,94,32)

x_hat = mean of x respect to axis (0,1,2) which is 10*94*94
c = x -x_hat

dc/dx = dc/dx + dc/dx_hat * dx_hat/hx
          =   1       +  (-1)          * 1/(10*94*94)
          =  0.999988683

Where was the mistake?

**Comment by ppwwyyxx** [comment]
`dc/dx_hat * dx_hat/dx`
None of these terms are real numbers. They are tensors.

**Comment by JaeDukSeo** [comment]
okay, my bad I wrote in a confusing way, it should be in a tensor as well.
Bu the general concept of taking derivative respect to the input (here x) is the same. 

Let me give you an example, we have a function 

def f(x): return tf.reduce_mean(x,(0,1,2,),keep_dims=True)

so the output would be a tensor shape of (1,1,1,32) right? 
But when we take the derivative respect to the original input x the derivative is 1/(10 * 94 * 94) 
For all variable in x - so it would be a tensor that has a shape of (10,94,94,32) but every value would be
0.00001131733

I believe you know how to do these by hand right Yuxin? 

I have attached some screenshots to help you understand what I mean. 

![image](https://user-images.githubusercontent.com/22832406/48329606-7ff1f600-e617-11e8-994a-4309de03ed7e.png)

![image](https://user-images.githubusercontent.com/22832406/48329616-884a3100-e617-11e8-8038-f84bd34d72e7.png)
